sycl/include/sycl/detail/defines_elementary.hpp

```diff
@@ -26,9 +26,16 @@
 #endif
 #endif // __SYCL_ALWAYS_INLINE
 
-#ifndef SYCL_EXTERNAL
+#ifdef SYCL_EXTERNAL
+#define __DPCPP_SYCL_EXTERNAL SYCL_EXTERNAL
+#else
+#ifdef __SYCL_DEVICE_ONLY__
+#define __DPCPP_SYCL_EXTERNAL __attribute__((sycl_device))
```

Comment:
Why wasn't this needed before? Was it because we would never encounter a path where `SYCL_EXTERNAL` wasn't defined by `__SYCL_DEVICE_ONLY__` was or was there a bug?

Reply:
Thanks for the review. Yeah, I think that's right. Here is the relevant code in the frontend before this change:

https://github.com/intel/llvm/blob/71c4b77c4e5d7af5668a69ff65e79644ad9e0141/clang/lib/Frontend/InitPreprocessor.cpp#L1299

```
  if (LangOpts.SYCLIsDevice) {
    Builder.defineMacro("__SYCL_DEVICE_ONLY__");
    Builder.defineMacro("SYCL_EXTERNAL", "__attribute__((sycl_device))");
...
```

So yeah if `__SYCL_DEVICE_ONLY__` was defined, `SYCL_EXTERNAL` would be defined to `__SYCL_DEVICE_ONLY__`

Reply:
Makes sense, thank you!